lib/classes/ConfigSchemaHandler.js

```diff
@@ -0,0 +1,124 @@
+    Object.freeze(this.schema.properties.layers);
+    Object.freeze(this.schema.properties.resources);
+  }
+
+  validateConfig(userConfig) {
+    const ajv = new Ajv();
+    const validate = ajv.compile(this.schema);
+    validate(userConfig);
+    if (validate.errors) {
+      const messages = errorMessageBuilder.buildErrorMessages(validate.errors);
+      for (const message of messages) {
+        switch (this.serverless.service.configValidationMode) {
+          case 'error':
+            throw new this.serverless.classes.Error(message);
+          case 'warn':
+            this.serverless.cli.log(message, 'Serverless', { color: 'red' });
+            break;
+          default:
+            throw new this.serverless.classes.Error(
+              '.configValidationMode should be equal to one of the allowed values `error` or `warn`'
+            );
+        }
+      }
+    }
+  }
+
+  defineTopLevelProperty(name, subSchema) {
+    this.schema.properties[name] = subSchema;
+  }
+
+  defineProvider(name, options = {}) {
+    // TODO: Remove this 'if' statement and place this piece of schema inside
+    // initial schema. This condition was added because some other tests
+    // were failing, see https://travis-ci.org/github/serverless/serverless/jobs/663153831
+    if (!this.schema.properties.provider.properties.name) {
+      this.schema.properties.provider.properties.name = { enum: [] };
+    }
+
+    this.schema.properties.provider.properties.name.enum.push(name);
+
+    if (options.provider) {
+      addPropertiesToSchema(this.schema.properties.provider, options.provider);
+    }
+
+    if (options.function && options.function.properties) {
+      addPropertiesToSchema(
+        this.schema.properties.functions.patternProperties[functionNamePattern],
+        options.function
+      );
+    }
+  }
+
+  defineCustomProperties(configSchemaParts) {
+    addPropertiesToSchema(this.schema.properties.custom, configSchemaParts);
+  }
+
+  defineFunctionEvent(name, configSchema) {
+    if (
+      !this.schema.properties.functions.patternProperties[functionNamePattern].properties.events
+        .items.anyOf
+    ) {
+      this.schema.properties.functions.patternProperties[
+        functionNamePattern
+      ].properties.events.items.anyOf = [];
+    }
+
+    this.schema.properties.functions.patternProperties[
+      functionNamePattern
+    ].properties.events.items.anyOf.push({
+      type: 'object',
+      properties: { [name]: configSchema },
+      required: [name],
+      additionalProperties: false,
+    });
+  }
+}
+
+function addPropertiesToSchema(subSchema, extenstion = { properties: {}, required: [] }) {
+  subSchema.properties = Object.assign(subSchema.properties, extenstion.properties);
+
+  if (!subSchema.required) subSchema.required = [];
+
+  if (Array.isArray(extenstion.required)) subSchema.required.push(...extenstion.required);
+
+  if (Object.prototype.hasOwnProperty.call(extenstion, 'additionalProperties')) {
```

Comment:
If we remove it, then the provider plugin would need to set it manually, as we agreed in https://github.com/serverless/serverless/pull/7335#discussion_r389319124. It might be a good choice because relaxing schema is a temporary mean

Reply:
Yes, I think we should treat it as temporary hack. So I would not support it here.

Ideally proposed schema in all cases should be complete

Reply:
Ok, I'll adjust it then

Reply:
Done in https://github.com/serverless/serverless/pull/7335/commits/430740bfa297dfc201378a87532cef1d3794c357